sw/device/lib/dif/dif_hmac.c

```diff
@@ -0,0 +1,221 @@
+         HMAC_STATUS_FIFO_DEPTH_MASK;
+}
+
+/**
+ * A helper funciton for calculating |HMAC_FIFO_MAX| - |get_fifo_entry_count()|.
+ */
+static uint32_t get_fifo_available_space(const dif_hmac_t *hmac) {
+  return HMAC_FIFO_MAX - get_fifo_entry_count(hmac);
+}
+
+dif_hmac_error_t dif_hmac_init(const dif_hmac_config_t *config,
+                               dif_hmac_t *hmac) {
+  if (config == NULL || hmac == NULL || config->base_addr.base == NULL ||
+      config->digest_endianness < kDifHmacBigEndian ||
+      config->digest_endianness > kDifHmacLittleEndian ||
+      config->message_endianness < kDifHmacBigEndian ||
+      config->message_endianness > kDifHmacLittleEndian ||
+      config->mode < kDifHmacModeHmac || config->mode > kDifHmacModeSha256) {
+    return kDifHmacInvalidArgument;
+  }
+
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, 0);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                      (1 << HMAC_INTR_ENABLE_HMAC_DONE) |
+                          (1 << HMAC_INTR_STATE_FIFO_FULL) |
+                          (1 << HMAC_INTR_STATE_HMAC_ERR));
+
+  // Set the byte-order of the input message.
+  uint32_t device_config = 0;
+  if (config->message_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_ENDIAN_SWAP);
+  }
+  // Set the byte-order of the digest.
+  if (config->digest_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_DIGEST_SWAP);
+  }
+
+  // Set HMAC engine to HMAC or SHA256-only mode.
+  if (config->mode == kDifHmacModeHmac) {
+    device_config |= (1 << HMAC_CFG_HMAC_EN);
+
+    // Set the key.
+    for (int i = 0; i < HMAC_KEY_REG_COUNT; i++) {
+      ptrdiff_t offset = i * sizeof(uint32_t);
+      mmio_region_write32(config->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                          config->key[i]);
+    }
+  }
+  device_config |= (1 << HMAC_CFG_SHA_EN);
+
+  // Write the configuration.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, device_config);
+
+  // Begin SHA or HMAC operation.
+  mmio_region_nonatomic_set_bit32(config->base_addr, HMAC_CMD_REG_OFFSET,
+                                  HMAC_CMD_HASH_START);
+
+  // Setup state tracking.
+  hmac->base_addr = config->base_addr;
+
+  return kDifHmacNoError;
+}
+
+dif_hmac_fifo_error_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
+                                         const void *data, size_t len,
+                                         size_t *bytes_sent) {
+  if (hmac == NULL || data == NULL) {
+    return kDifHmacFifoInvalidArgument;
+  }
+
+  const uint8_t *data_sent = (uint8_t *)data;
+  int words_remaining = len / sizeof(uint32_t);
+  int bytes_remaining = len % sizeof(uint32_t);
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
```

Comment:
Why not add a third loop to bring `data_sent` into `uint32_t` alignment, and then use `read_32` from memory.h for the central loop?

Reply:
Are you suggesting writing individual bytes until `data_sent` is aligned?

Reply:
I don't think so. Let's say `len` is 7 bytes, and aligned as such
```
XXXXXX01 23456789 ABCDXXXX
      ^
  data_sent
```
Since `data_sent` ins't necessarily aligned to begin with I think it would require three loops to

1. Bring `data_sent` into alignment by writing bytes from the first, unaligned word `{ 0x01 }`
```
XXXXXX01 23456789 ABCDXXXX
         ^
     data_sent
```
2. Loop on aligned words with `read_32` to send all 32-bit words `{ 0x23456789 }`
```
XXXXXX01 23456789 ABCDXXXX
                  ^
              data_sent
```
3. Write the final bytes from the last word `{ 0xAB, 0xCD }`

@mcy is this what you're suggesting?

Reply:
(Moved my reply over email here)

Yes, you’re right. For some reason I only thought about the beginning not the end. Sorry for the confusion!

What do you think about the following snippet? Please feel free to ignore this, change constants to `sizeof`s, etc.. I know this goes against @silvestrst's earlier comment but somehow looks simpler to me.

```
const uint8_t *data_sent = (const uint8_t *)data;
int rem_bytes = len;

while(rem_bytes > 0 && get_fifo_available_space(hmac) > 0) {
  if ((data_sent & 0x3u) == 0 && rem_bytes >= 4) {
    // Word aligned
    uint32_t word = read_32(data_sent);
    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
    data_sent += 4;
    rem_bytes -= 4;
  } else {
    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
    ++data_sent;
    --rem_bytes;
  }
}
```

Reply:
I do like the idea of having a single loop, and I like how it reduces code duplication by combining the word alignment and trailing byte writes into a single case. I think the way it is now, with the three independent loops, is easier to reason about since the three steps always happen in the same order and the code makes that clear.